src/main/java/org/prebid/server/settings/S3ApplicationSettings.java

```diff
@@ -0,0 +1,219 @@
+                        )
+                )
+                .map(fileContent -> jacksonMapper.decodeValue(fileContent, Account.class))
+                .flatMap(account -> {
+                    if (!Objects.equals(account.getId(), accountId)) {
+                        return Future.failedFuture(new PreBidException(
+                                "Account with id %s does not match id %s in file".formatted(accountId, account.getId()))
+                        );
+                    }
+                    return Future.succeededFuture(account);
+                })
+                .recover(ex -> {
+                    if (ex instanceof DecodeException) {
+                        return Future
+                                .failedFuture(
+                                        new PreBidException(
+                                                "Invalid json for account with id %s".formatted(accountId)));
+                    }
+                    // if a previous validation already yielded a PreBidException, just return it
+                    if (ex instanceof PreBidException) {
+                        return Future.failedFuture(ex);
+                    }
+                    return Future
+                            .failedFuture(new PreBidException("Account with id %s not found".formatted(accountId)));
+                });
+    }
+
+    @Override
+    public Future<StoredDataResult> getStoredData(
+            String accountId,
+            Set<String> requestIds,
+            Set<String> impIds,
+            Timeout timeout) {
+
+        return getFileContents(storedRequestsDirectory, requestIds)
+                .compose(storedIdToRequest -> getFileContents(storedImpressionsDirectory, impIds)
+                        .map(storedIdToImp ->
+                                buildStoredDataResult(storedIdToRequest, storedIdToImp, requestIds, impIds))
+                );
+    }
+
+    private StoredDataResult buildStoredDataResult(
+            Map<String, String> storedIdToRequest,
+            Map<String, String> storedIdToImp,
+            Set<String> requestIds,
+            Set<String> impIds
+    ) {
+
+        final List<String> missingStoredRequestIds =
+                getMissingStoredDataIds(storedIdToRequest, requestIds).stream()
+                        .map("No stored request found for id: %s"::formatted).toList();
+        final List<String> missingStoredImpressionIds =
+                getMissingStoredDataIds(storedIdToImp, impIds).stream()
+                        .map("No stored impression found for id: %s"::formatted).toList();
+
+        return StoredDataResult.of(
+                storedIdToRequest,
+                storedIdToImp,
+                Stream.concat(
+                        missingStoredImpressionIds.stream(),
+                        missingStoredRequestIds.stream()).toList());
+    }
+
+    private List<String> getMissingStoredDataIds(Map<String, String> fileContents, Set<String> responseIds) {
```

Comment:
Why List when you can use Set directly? So intermediate conversions can be omitted (I am talking about `.stream().toList()` calls that create redundant objects).